ansible_base/authentication/utils/authentication.py

```diff
@@ -0,0 +1,120 @@
+        new_username = exact_match[0].user.username
+        logger.info(f"Authenticator {authenticator.name} already authenticated {uid} as {new_username}")
+        return new_username
+
+    # We didn't get an exact match. If any other provider is using this uid our id will be uid<hash>
+    if AuthenticatorUser.objects.filter(uid=uid).count() != 0:
+        # Some other provider is providing this username so we need to create our own username
+        new_username = get_local_username({'username': uid})
+        logger.info(
+            f'Authenticator {authenticator.name} wants to authenticate {uid} but that'
+            f'username is already in use by another authenticator,'
+            f' the user from this authenticator will be {new_username}'
+        )
+        return new_username
+
+    # We didn't have an exact match but no other provider is servicing this uid so lets return that for usage
+    logger.info(f"Authenticator {authenticator.name} is is able to authenticate user {uid} as {uid}")
+    return uid
+
+
+def get_or_create_authenticator_user(
+    username: str, authenticator: Authenticator, user_details: dict = {}, extra_data: dict = {}
+) -> Tuple[AbstractUser, Optional[AuthenticatorUser], bool]:
+    """
+    Create the user object in the database along with it's associated AuthenticatorUser class.
+    In some cases, the user may already be created in the database.
+    This assumes that the given user_name has already been resolved using the determine_username_from_uid but will do a quick double check
+
+    This should be called any non-social auth plugins.
+
+    Inputs
+    username: The username we are going to be created
+    user_details: Any details about the user from the source (first name, last name, email, etc)
+    authenticator: The authenticator authenticating the user
+    extra_data: Any additional information about the user provided by the source.
+                For example, LDAP might return sn, location, phone_number, etc
+    """
+
+    created = None
+    try:
+        # First see if we have an auth user and if so update it
+        auth_user = AuthenticatorUser.objects.get(uid=username, provider=authenticator)
+        auth_user.extra_data = extra_data
+        auth_user.save()
+        created = False
+    except AuthenticatorUser.DoesNotExist:
+        # Ensure that this username is not already tired to another authenticator
+        auth_user = AuthenticatorUser.objects.filter(uid=username).first()
+        if auth_user is not None:
+            logger.error(
+                f'Authenticator {authenticator.name} attempted to create an AuthenticatorUser for {username}'
+                f' but that id is already tied to authenticator {auth_user.provider.name}'
+            )
+            return None, None, None
+
+    # ensure the authenticator isn't trying to pass along a cheeky is_superuser in user_details
+    allowed_keys = ["first_name", "last_name", "email"]
```

Comment:
If that's a constant, better to move it to module level and name appropriately

See https://peps.python.org/pep-0008/#constants